proposals/0000-ghc-module-naming.rst

```diff
@@ -0,0 +1,139 @@
+
+  * These libraries come with no stability guarantees: they may change at short notice.
+
+In addition we already have:
+
+* ``ghc``: this library exposes GHC as a library, through the (currently ill-defined) GHC API.
+
+All these libraries follow the Haskell Package Versioning Policy (PVP).
+
+The question arises of *what module names should be used*. For example, suppose that all three exposed a module called ``Data.Tuple``.  In principle that would be fine -- GHC allows you
+to use the package name in the ``import`` statement, to disambiguate.  But it's *extremely* confusing.  This proposal articulates a set of conventions to
+help us design module names.
+
+The proposal
+============
+
+This proposal is split into four sub-proposals for easier discussion.  Each sub-proposal builds on the
+earlier ones -- they are increments, not alternatives.
+
+The goals of this proposal are deliberately limited to establish naming conventions.  We do not propose
+any changes to ``ghc`` or to ``cabal``.
+
+Proposal 1
+-----------
+
+* Modules in ``base``, ``ghc-experimental``, ``ghc-prim``, ``ghc-internals`` etc should all have distinct names.
+
+That principle leads immediately to the question: what should those names be?  Hence proposal 2.
+
+Proposal 2
+-----------
+
+* Modules in GHC's internal libraries (``ghc-prim``, ``ghc-internals`` etc) should be of form ``GHC.*``.
+* Modules in ``ghc-experimental`` should be of form ``Experimental.*``.
+* Modules in ``base`` should not have either of these prefixes.
+
+So example we might have
+
+* ``GHC.Tuple`` in ``ghc-internals``,
+* ``Experimental.Tuple`` or ``Experimental.Data.Tuple`` in ``ghc-experimental``
+* ``Data.Tuple`` in ``base``
+
+Proposal 3
+-----------
+
+The current ``base`` API exposes many modules starting with ``GHC.*``, so the proposed conventions could only
+apply to *new* modules.
+
+* Over time, and only with the agreement and support of the Core Libraries Committee, we may remove some ``GHC.*`` modules
+  from ``base``, especially ones that are barely used, or are manifestly "internal" (i.e. part of the implementation
+  of other, more public functions).
+  Of course there would be a significant deprecation cycle, to allow client libraries to adapt.
+
+Proposal 3 only expresses a direction of travel.  We will have to see what the CLC's attitude is,
+and what the Haskell community thinks.  Anything that disturbs the API of base needs to be considered
+rather carefully.
+
+
+Proposal 4
+------------
+
+* The public API of package ``ghc`` (GHC as a library) should have modules of form ``GhcAPI.*``.
+
+All of the modules in package ``ghc`` currently start with ``GHC.*`` which correctly signals that they are part of GHC's internals.
+As part of the GHC API redesign (a HF project in its own right, currently stalled) it would be very helpfult
+to modules with stable APIs, and a new prefix, such as ``GhcAPI.*``.
```

Comment:
I am not so sure about this. bifurcating GHC into "public API" vs "internals" is just one way to go about organizing its code (and I personally don't think it is a good one). Regardless of whether it is good or bad, the module naming I think should be independent of such matter of internal organization: modules in GHC should be named the same way whether they are more or less stable.

(If we really had a stable deivions to make, we would have a separate library per https://nikita-volkov.github.io/internal-convention-is-a-mistake/, but we are nowhere near that.)

"API" is also, I think, bad bargain, the "application" and "programming" don't mean anything in particular, and "interface" is quite redundant when all module naming is for interface purposes.

------

`GHC.` I still think is the best name for the compiler proper. GHC stands for Glasgow Haskell *Compiler* --- it's in the name.

The counterparts to `ghc-internals`/`ghc-prim` are called things like `libgcc` (GCC) or `libcompiler-rt` (Clang). The `lib` prefix of C/C++ libraries is doing some work here, `compiler-rt` is rather generic. but I think the `rt` is good. We have "RTS" for just the C part, but it might almost be better to think of the runtime as something encompassing all of `rts` + `ghc-prim` + `ghc-internals`. None of this stuff is part of the compiler proper, but all of it is unstable "support code" propping up the code the user actually wrote. And functionality does in fact move between the various parts of it with some fluidity --- for example @dcoutts is working on moving some IO stuff *back* into C for the threaded runtime (like the unthreaded runtime does today) for more performance [`io_uring`](https://en.wikipedia.org/wiki/Io_uring) support.

So I don't yet have an alternative name I really like, but given the above, maybe something like `GHR` for "Glasgow Haskell Runtime" would make sense.

Reply:
"bifurcating GHC into "public API" vs "internals" is just one way to go about organizing its code" -- I think/hope that the ghc api proposal is _not_ this. Rather, it is to create a _new_ set of modules with _new_ functions that promise to be a stable (and usable) api, and which are _not_ intended for internal use by ghc. That is to say, ghc itself does _not_ dogfood the ghc api (although ghci may choose to do so) and instead the functions provided by the api are designed from the start for external consumption, perhaps glossing over certain tricky but rare bits at first, especially ones highly dependent on less stable datatypes. As such, having a distinct namespace of some sort for the API is very important regardless.

I can imagine such a package perhaps evolving over time to be a distinct package on top of ghc, but that seems orthogonal to this discussion. And, well, now that I think about it, since the API proposal currently is stalled, we probably need not worry about it at all in this discussion -- I just wanted to clarify what the idea is, last I understood it, for the sake of future discussion.

Reply:
> "bifurcating GHC into "public API" vs "internals" is just one way to go about organizing its code" -- I think/hope that the ghc api proposal is not this. Rather, it is to create a new set of modules with new functions that promise to be a stable (and usable) api, and which are not intended for internal use by ghc. That is to say, ghc itself does not dogfood the ghc api (although ghci may choose to do so) and instead the functions provided by the api are designed from the start for external consumption, perhaps glossing over certain tricky but rare bits at first, especially ones highly dependent on less stable datatypes. As such, having a distinct namespace of some sort for the API is very important regardless.

Yes, @gbaz has it exactly right.   `GhcAPI.*` is a shim around the internal modules `GHC.*`, one that has stronger stability guarantees.  But those internal modules *must* remain available because we can't predict everything that a client of the `ghc` library may want to do.  I hope that if someone finds they can only do something through `GHC.*` they will petition the GHC API working group (still in the womb) to add a suitable function to `GhcAPI.*`.

We must be able to reorganise and refactor GHC's internals without constraint.  At the moment we simply don't know what clients of `ghc` are relying on, so we may well mess up their lives without ever knowing.  If we had `GhcAPI.*` we'd know which modules we needed to take (much) more care with.

The details are not important;  I'm just using this proposal as a way establish, in principle. a namespace for a stable GHC API.

Reply:
I like ignoring it for now; I would like it if this proposal didn't mention that at all, and if/when we do that we worry about its module names then.

That brings us back to using `GHC.*` both for the compiler (internals) and for `ghc-internals`/`ghc-prim`. In another thread, this was brought up as a deficiency. This proposal doesn't yet talk about this problem, but shouldn't it?